itest/rust/src/builtin_tests/containers/signal_test.rs

```diff
@@ -349,6 +350,97 @@ fn signal_symbols_engine_inherited_internal() {
     node.free();
 }
 
+// Test that signal API methods ending with `_engine` accept engine types as receivers.
+#[cfg(since_api = "4.2")]
+#[itest]
+fn signal_symbols_connect_engine() {
+    // No tree needed; signal is emitted manually.
+    let mut node = Emitter::new_alloc();
+    let mut engine = Node::new_alloc();
+    engine.set_name("hello");
+
+    node.signals()
+        .property_list_changed()
+        .connect_engine(&engine, |this| {
+            assert_eq!(this.get_name(), StringName::from("hello"));
+        });
+
+    node.signals()
+        .property_list_changed()
+        .connect_builder()
+        .object(&engine)
+        .method_engine(|this| {
+            assert_eq!(this.get_name(), StringName::from("hello"));
+        })
+        .done();
+
+    node.signals().property_list_changed().emit();
+
+    node.free();
+    engine.free();
+}
+
+// Test that rustc is capable of inferring the parameter types of closures passed to the signal API's connect methods.
+#[cfg(since_api = "4.2")]
+#[itest]
+fn signal_symbols_connect_inferred() {
+    let mut user = Emitter::new_alloc();
+    let mut engine = Node::new_alloc();
+
+    user.signals()
+        .child_entered_tree()
+        .connect_engine(&engine, |this, mut child| {
+            // Use methods that `Node` declares
+            let _ = this.get_path(); // ref
+            this.set_unique_name_in_owner(true); // mut
+
+            // `child` is also a node
+            let _ = child.get_path(); // ref
+            child.set_unique_name_in_owner(true); // mut
+        });
```

Comment:
Just for my understanding, where do you see the main advantage over `connect()`  here?

Wouldn't the following work (if `engine` is no longer used -- otherwise with an extra clone)?
```suggestion
    let mut user = Emitter::new_alloc();
    let mut engine = Node::new_alloc();

    user.signals()
        .child_entered_tree()
        .connect(move |mut child| {
            // Use methods that `Node` declares
            let _ = engine.get_path(); // ref
            engine.set_unique_name_in_owner(true); // mut

            // `child` is also a node
            let _ = child.get_path(); // ref
            child.set_unique_name_in_owner(true); // mut
        });
```

Reply:
Also, this gives minor ergonomic benefits when a single object is use (no need to copy or have `move` closure), but you'd still need this pattern if more than 1 `Gd` needs to be accessed from within the closure.

In comparison, `connect_obj` exists because it does the heavy lifting of rebinding the object so that the user can safely access it, especially when `self.signals().xy().emit()` is called. This isn't easily possible with other existing methods.

Reply:
It's a minor advantage, but Gd-Rust already has enough boilerplate, and each line removed counts towards reducing it.

Moving engine `Gds` into the closures is easy, but if you want to reuse said GD elsewhere, you need to add a line where you explicitly create a new variable, cloning the instance you want to reuse:
```rust
let user_class = Emitter::new_alloc();
let mut some_node = Node::new_alloc();

let node_clone = some_node.clone(); // extra line
user_class.signals().ready().connect(move || {
    node_clone.do_something();
});

// keep using `some_node`
```

Compare connecting to a signal in GDScript/C# to Rust:
```go
var bound_obj = ...
emitter.signal_name.connect(func(): do_something(bound_obj))
```

```csharp
var bound_obj = ...;
emitter.signal_name += () => do_something(bound_obj);
```

```rust
let source_obj = ...;
let bound_obj = source_obj.clone();
emitter.signals().signal_name().connect(move || { do_something(bound_obj) });
```

So in my opinion, any boilerplate reduction is welcome.

---

The bellow is not directly related to the issue above, consider it as food for thought:

The most common use case in my experience is connecting to a signal that's on a exported field of your class, look at how many extra lines this takes:

```rust
#[derive(GodotClass)]
#[class(base = Control, init)]
pub struct SettingsUI {
    base: Base<Control>,
    #[init(node = "close_menu_button")]
    close_menu_button: OnReady<Gd<BaseButton>>,
}

#[godot_api]
impl IControl for SettingsUI {
    fn ready(&mut self) {
        self.close_menu_button
            .clone()                      // needed due to borrow checker, since `signals` takes `&mut self`, even though internally it doesn't really need a mutable reference
            .signals()                   // extra call other languages don't have
            .pressed()
            .connect_builder()   // extra call
            .object_self(self)      // extra call
            .method_mut(|this| this.base_mut().hide())
            .flags(ConnectFlags::DEFERRED) // connecting deferred is almost always preferable, so I just do it by default. 
            .done();                      // extra call
    }
}
```

In my opinion, the API should be (and we can!) changed to allow the following:

```rust
#[godot_api]
impl IControl for SettingsUI {
    fn ready(&mut self) {
        self.close_menu_button
            .signals()         // we could get rid of this call, but for now, it's fine.
            .pressed()       // make this take `&self` instead of `&mut self`: no borrow checker issues, functionality remains the same.
            .connect_other_deferred(self, |this| this.base_mut().hide());
    }
```

We get users having issues with double bind almost every day, which are easily solved by
connecting deferred. I believe that should be incentivized and facilitated.

Reply:
Thanks for the detailed elaboration!

Agree on unnecessary clone, and we can possibly collapse `object_self` + `method_mut` in the builder as well. The `ToSignalObj` trait abstracts over `&mut WithUserSignals` and `Gd<T>` and should (hopefully) be able to provide some groundwork here.

I don't agree with `deferred`, at least not right now.
- First, I think we should see if we can establish patterns that make borrow errors less frequent. `emit()` for example is carefully crafted so it can be re-entrant _on the same object_.
- Deferred isn't the solution to everything. It introduces asynchronous execution, and sometimes you want effects to be applied immediately, because later logic in the same frame depends on it.
- The approach of promoting everything to the top level still causes combinatoric explosion, even if used frequently. Some people use one-shot signals a lot. For some it's important to have cross-thread signals due to their architecture.

Not saying we can never do this, but these are very clearly QoL features that can possibly be added later, after more careful evaluation.

---

Rather than supporting every use case in the shortest possible syntax, we should make the builder modular, so that people can write their own abstractions on top. Imagine you could write extension traits like this:
```rs
#[godot_api]
impl IControl for SettingsUI {
    fn ready(&mut self) {
        self.close_menu_button
            .signals() 
            .pressed()
            .connect_builder()
            .my_custom_connect(|this| this.base_mut().hide());
    }
```
but this might get much harder with the macro-based approach now.

Reply:
> Deferred isn't the solution to everything. It introduces asynchronous execution, and sometimes you want effects to be applied immediately, because later logic in the same frame depends on it.

I believe it is for 90% of cases, which has been (anecdotally) proven by the help requests we get on discord.
For the cases were a user wants to apply something immediately, they can just not connect with the `Deferred` flag and handle the consequences on their own.

When communication needs to happen immediately, there's a good chance that you're better off communicating directly in Rust rather than going through the Godot layer (again, *in my experience*).

> Rather than supporting every use case in the shortest possible syntax

I don't think we should ever support every use case in the shortest possible syntax, my argument is that `connect_deferred` is more common than a regular `connect`, and thus it should be supported.

Reply:
> I believe it is for 90% of cases, which has been (anecdotally) proven by the help requests we get on discord.

With the same argument, you could justify `call_deferred` everywhere because people constantly run into borrow issues with `Gd::bind/bind_mut`. And then it turns out that most of them can be solved by proper use of `base/base_mut`. And we can probably do better in conveying these concepts.

Deferred signals as a result of borrow errors _are_ a workaround because of a Rust-specific problem, they are not the default in Godot itself. While I don't think we can 100% solve every case, I still believe more effort should be put into promoting safe usage of non-deferred signals, both with API design and documentation.